kernel: move thread definitions to thread.h

nashif opened the pull request:
Move thread definitions to its own header to avoid redeclaration and
redefinition of types which is not allowed in some standards.

nashif commented:
cleaner solution mentioned in #2853, should remove the compiler option once this go in.

tejlmand commented:
> cleaner solution mentioned in #2853, should remove the compiler option once this go in.

Should probably be: #28531

tejlmand reviewed (approved):
lgtm

andrewboie commented on include/thread.h:
```diff
@@ -0,0 +1,254 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_THREAD_H_
```
suggestion: put all this in include/sys/thread.h, to at least space this under sys/ in the include space rather than top-level. I think we want to avoid any new toplevel headers.

andrewboie commented on include/thread.h:
```diff
@@ -0,0 +1,254 @@
+	struct _thread_arch arch;
+};
+
+typedef struct k_thread _thread_t;
```
should probably purge any code still using _thread_t

tyrantlizard77 reviewed (commented):
A good start but I'd like this in a sub-folder of the include space, perhaps include/sys/thread.h? instead of at top-level.

eventually I'd like to see the thread APIs also placed in this header, not just the struct declarations, but that can be another PR

ioannisg reviewed (approved):
Looks good (and agree with @andrewboie 's suggestion)

nashif commented:
> A good start but I'd like this in a sub-folder of the include space, perhaps include/sys/thread.h? instead of at top-level.

I am fine with some other location, however, I really want to see this following some pattern and we should move other top level headers hanging in include/ like kernel_includes.h, kernel_structs.h, etc.)

I think it is time for the yearly review of include/ and where headers should be, it is been a while and things are inconsistent again.

@carlescufi can you please put this on the agenda for the next API meeting?

nashif committed:
kernel: move thread definitions to thread.h

Move thread definitions to its own header to avoid redeclaration and
redefinition of types which is not allowed in some standards.

Fixes #29937

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nandojve reviewed (approved):
Hi @nashif 

I already create a PR to remove that kernel include from PSoC-6. It seems the last one.
https://github.com/zephyrproject-rtos/zephyr/pull/29459/commits/546f8dd85e2d3c961bd886af6ec4cc4944af3a30
Fell free to incorporate.

tyrantlizard77 reviewed (commented):
You are moving kernel_structs.h and kernel_version.h where it resides in the public include path. I think we need to put redirection headers where they used to be which emit a warning, to not completely break end users.
